### PR TITLE
Feature/custom error messages real

### PR DESCRIPTION
**Pull request by aaronjameslang**
#2

### CODE REVIEW

**Comment by aaronjameslang** [comment]
This is WIP

**Comment by aaronjameslang** [comment]
@i-am-tom This is ready for consideration now :slightly_smiling_face:

File: README.md [hunk 1]
```diff
@@ -135,6 +135,8 @@ Schemer\Formatter::text()
 
 ### `Schemer\Validator`
 
+Some methods take an optional custom error `string $error`
```
Comment (i-am-tom) [inline README.md]: My only issue is that I'd probably pad this out a bit. "For many of these methods, a custom `$error` string can be supplied as the final parameter to override the Schemer default" or something like that. Really just a preference thing... sounds more... flamboyant :star:

File: test/Validator/Real.php [hunk 2]
```diff
@@ -70,6 +96,15 @@
                     ->errors()
             )->toBe(['not at most 2']);
         });
+
+        it('rejects higher floats, with custom error', function () {
+            expect(
+                (new Real)
+                    ->max(9000, "It's over 9000!")
```
Comment (i-am-tom) [inline test/Validator/Real.php:103]: :zap:

**Comment by imjoehaines** [comment]
So I haven't been able to think of a clean way to do it, but I was thinking more along the lines of tokenising the error messages from the validation functions and having a i18n-like process convert those into error messages en masse, so you only need to overwrite each message once. It's a bit less flexible but I don't think I'd ever want to have a different message for the same validation on two different fields. e.g. if I had this (real code):

``` php
$this->validator = V::assoc([
    'title' => V::text()->min(10)->max(50),
    'summary' => V::text()->min(25)->max(500),
    'description' => V::text()->min(50)->max(3200),
]);
```

I wouldn't want to have to overwrite the `text` errors multiple times;

``` php
$this->validator = V::assoc([
    'title' => V::text('not text')->min(10, 'not long enough')->max(50, 'not short enough'),
    'summary' => V::text('not text')->min(25, 'not long enough')->max(500, 'not short enough'),
    'description' => V::text('not text')->min(50, 'not long enough')->max(3200, 'not short enough'),
]);
```

Something along the lines of:

``` php
$this->validator = V::assoc([
    'title' => V::text()->min(10)->max(50),
    'summary' => V::text()->min(25)->max(500),
    'description' => V::text()->min(50)->max(3200),
])->translate([
    V::NOT_STRING => 'not text',
    V::TOO_SHORT => 'not long enough',
    V::TOO_LONG => 'not short enough',
]);
```

would make more sense to me (with some kind of naïve method to put the array key in the message). I'm not sure if there's anywhere in the current code that would naturally be responsible for doing this kind of thing, so with my annoying person using this library hat on; make your code worse so mine can be better kthx

Buuut doing it this way is a bit more flexible and I can stick the custom messages in variables so it's not the end of the world, either way works.

**Comment by i-am-tom** [comment]
No, no, I see what you mean. Presumably, any nested structures would also need the `->translate` method call, too? Obviously, the usual i18n approach wasn't my favourite because it's inherently stateful, but I kinda like the idea of translating validators - I think of the library as a set of lenses for data, and that'd be lenses for those lenses, which is so meta/10.

**Comment by aaronjameslang** [comment]
@i-am-tom How's that? I can put exactly what you suggested if you want, but I think this says the same thing in fewer words.

**Comment by i-am-tom** [comment]
It's a full stop away from perfect <3

File: README.md [hunk 3]
```diff
@@ -135,6 +135,8 @@ Schemer\Formatter::text()
 
 ### `Schemer\Validator`
 
+Many of these methods accept an optional custom `$error` string, which overrides the default error
```
Comment (i-am-tom) [inline README.md]: .

**Comment by i-am-tom** [comment]
<3 <3 <3

**Comment by aaronjameslang** [comment]
:D

**Pull request merged**